compiler-rt/lib/asan/asan_poisoning.cpp

```diff
@@ -20,11 +20,70 @@
 #include "sanitizer_common/sanitizer_flags.h"
 #include "sanitizer_common/sanitizer_interface_internal.h"
 #include "sanitizer_common/sanitizer_libc.h"
+#include "sanitizer_common/sanitizer_ring_buffer.h"
+#include "sanitizer_common/sanitizer_stackdepot.h"
 
 namespace __asan {
 
+using PoisonRecordRingBuffer = RingBuffer<struct PoisonRecord>;
+
 static atomic_uint8_t can_poison_memory;
 
+static Mutex poison_records_mutex;
+static PoisonRecordRingBuffer *poison_records = nullptr;
+
+void AddPoisonRecord(const PoisonRecord &new_record) {
+  if (flags()->poison_history_size <= 0)
+    return;
+
+  poison_records_mutex.Lock();
+
+  if (poison_records == nullptr)
+    poison_records = PoisonRecordRingBuffer::New(flags()->poison_history_size);
+
+  poison_records->push(new_record);
+
+  poison_records_mutex.Unlock();
+}
+
+bool FindPoisonRecord(uptr addr, const PoisonRecord &match) {
+  if (flags()->poison_history_size <= 0)
+    return false;
+
+  poison_records_mutex.Lock();
+
+  if (poison_records) {
+    for (unsigned int i = 0; i < poison_records->size(); i++) {
+      struct PoisonRecord record = (*poison_records)[i];
+      if (record.begin <= addr && addr < record.end) {
+        internal_memcpy((void *)&match, (void *)&record,
+                        sizeof(struct PoisonRecord));
+        poison_records_mutex.Unlock();
+        return true;
+      }
+    }
+  }
+
+  poison_records_mutex.Unlock();
+  return false;
+}
+
+__attribute__((no_thread_safety_analysis)) void SANITIZER_ACQUIRE(
```

Comment:
why `__attribute__((no_thread_safety_analysis))` ?

Reply:
The reviewed version of the code had:
```
 if (flags()->poison_history_size <= 0)
   return; // Did not lock
```

Thread safety analysis requires that every code path acquires the lock.

Since I've removed the if statement per your feedback below, I've also removed `__attribute__((no_thread_safety_analysis))`.